File: framec_tests/javascript/output/simple_handler_calls.js
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

const FrameEvent = require("../framelang/FrameEvent")

class SimpleHandlerCalls {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sInit_;
        this.#compartment = new SimpleHandlerCallsCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    A() {
        let e = FrameEvent("A",null);
        this.#mux(e);
    }
    
    B() {
        let e = FrameEvent("B",null);
        this.#mux(e);
    }
    
    C() {
        let e = FrameEvent("C",null);
        this.#mux(e);
    }
    
    D() {
        let e = FrameEvent("D",null);
        this.#mux(e);
    }
    
    E() {
        let e = FrameEvent("E",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sInit_:
                this.#sInit_(e);
                break;
            case this.#sA_:
                this.#sA_(e);
                break;
            case this.#sB_:
                this.#sB_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sInit_(e) {
        switch (e._message) {
            case "A":
                {
                let compartment =  new SimpleHandlerCallsCompartment(this.#sA_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "B":
                {
                let compartment =  new SimpleHandlerCallsCompartment(this.#sB_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "C":
                {
                this.A();
                return;
                
                return;
                }
                
            case "D":
                {
                this.B();
                return;
                let compartment =  new SimpleHandlerCallsCompartment(this.#sA_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "E":
                {
                this.D();
                return;
                this.C();
                return;
                
                return;
                }
                
        }
    }
    
    #sA_(e) {
        switch (e._message) {
        }
    }
    
    #sB_(e) {
        switch (e._message) {
        }
    }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};

//=============== Compartment ==============//

class SimpleHandlerCallsCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class SimpleHandlerCallsController extends SimpleHandlerCalls {

	constructor() {
	  super()
	}
};

********************/

File: framec_tests/csharp/SimpleHandlerCalls/SimpleHandlerCalls.cs
using FrameLang;
#nullable disable
namespace SimpleHandlerCalls
{

class SimpleHandlerCalls
{

    private SimpleHandlerCallsCompartment _compartment_;
    private SimpleHandlerCallsCompartment _nextCompartment_;



    public SimpleHandlerCalls(){


        // Create and intialize start state compartment.

        this._state_ = (int)SimpleHandlerCallsState.INIT;
        this._compartment_ = new SimpleHandlerCallsCompartment(this._state_);
        this._nextCompartment_ = null;


        // Send system start event
        FrameEvent frameEvent = new FrameEvent(">", null);
        this._mux_(frameEvent);

    }

    // states enum
    private enum SimpleHandlerCallsState
    {

        INIT = 0,
        A = 1,
        B = 2
    }
    //====================== Multiplexer ====================//

    private void _mux_(FrameEvent e)
    {
        switch (this._compartment_.state)
        {
            case (int)SimpleHandlerCallsState.INIT:
                this._sInit_(e);
                break;
            case (int)SimpleHandlerCallsState.A:
                this._sA_(e);
                break;
            case (int)SimpleHandlerCallsState.B:
                this._sB_(e);
                break;
        }

        if( this._nextCompartment_ != null)
        {
            var nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if (nextCompartment._forwardEvent != null &&
               nextCompartment._forwardEvent._message == ">")
            {
                this._mux_(new FrameEvent( "<", this._compartment_.ExitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            }
            else
            {
                this._doTransition_(nextCompartment);
                if (nextCompartment._forwardEvent != null)
                {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }

    //===================== Interface Block ===================//

    public void A() {
        FrameEvent e = new FrameEvent("A",null);
        this._mux_(e);
    }

    public void B() {
        FrameEvent e = new FrameEvent("B",null);
        this._mux_(e);
    }

    public void C() {
        FrameEvent e = new FrameEvent("C",null);
        this._mux_(e);
    }

    public void D() {
        FrameEvent e = new FrameEvent("D",null);
        this._mux_(e);
    }

    public void E() {
        FrameEvent e = new FrameEvent("E",null);
        this._mux_(e);
    }


    //===================== Machine Block ===================//

    SimpleHandlerCallsCompartment compartment;


    private void _sInit_(FrameEvent e)
    {
        if (e._message == "A")
        {


            compartment =  new SimpleHandlerCallsCompartment((int)SimpleHandlerCallsState.A);


            this._transition_(compartment);

            return;
        }
        else if (e._message == "B")
        {


            compartment =  new SimpleHandlerCallsCompartment((int)SimpleHandlerCallsState.B);


            this._transition_(compartment);

            return;
        }
        else if (e._message == "C")
        {
            A();
            return;

            return;
        }
        else if (e._message == "D")
        {
            B();
            return;


            compartment =  new SimpleHandlerCallsCompartment((int)SimpleHandlerCallsState.A);


            this._transition_(compartment);

            return;
        }
        else if (e._message == "E")
        {
            D();
            return;
            C();
            return;

            return;
        }
    }

    private void _sA_(FrameEvent e)
    {
    }

    private void _sB_(FrameEvent e)
    {
    }


    //=============== Machinery and Mechanisms ==============//

    private int _state_;

    private void _transition_(SimpleHandlerCallsCompartment compartment)
    {
        this._nextCompartment_ = compartment;
    }

    private void _doTransition_(SimpleHandlerCallsCompartment nextCompartment)
    {
        this._mux_(new FrameEvent("<", this._compartment_.ExitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.EnterArgs));
    }

    public string state_info(){
        return this._compartment_.state.ToString();
        }

}

//=============== Compartment ==============//

class SimpleHandlerCallsCompartment
{

    public int state;

    public SimpleHandlerCallsCompartment(int state)
    {
        this.state = state;
    }

    public Dictionary<string, object> StateArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> StateVars { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> EnterArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> ExitArgs { get; set; } = new Dictionary<string, object>();
    public FrameEvent _forwardEvent = new FrameEvent();
}


/********************

class SimpleHandlerCallsController : SimpleHandlerCalls
{
        public SimpleHandlerCallsController() : base()
        {
        }
}

********************/
}